src/hypertable.c

```diff
@@ -849,6 +853,160 @@ hypertable_create_schema(const char *schema_name)
+ * are made unique by appending the trigger OID, which we do not
+ * know. Instead, we have to search all triggers.
+ */
+static Oid
+get_insert_blocker_trigger(Oid relid)
+{
+	Relation	tgrel;
+	ScanKeyData skey[1];
+	SysScanDesc tgscan;
+	HeapTuple	tuple;
+	Oid			tgoid = InvalidOid;
+
+	tgrel = heap_open(TriggerRelationId, AccessShareLock);
+
+	ScanKeyInit(&skey[0],
+				Anum_pg_trigger_tgrelid,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(relid));
+
+	tgscan = systable_beginscan(tgrel, TriggerRelidNameIndexId, true,
+								NULL, 1, skey);
+
+	while (HeapTupleIsValid(tuple = systable_getnext(tgscan)))
+	{
+		Form_pg_trigger trig = (Form_pg_trigger) GETSTRUCT(tuple);
+
+		if (TRIGGER_TYPE_MATCHES(trig->tgtype, TRIGGER_TYPE_ROW, TRIGGER_TYPE_BEFORE, TRIGGER_TYPE_INSERT) &&
+			strncmp(INSERT_BLOCKER_NAME, NameStr(trig->tgname), strlen(INSERT_BLOCKER_NAME)) == 0 &&
+			trig->tgisinternal)
+		{
+			tgoid = HeapTupleGetOid(tuple);
+			break;
+		}
+	}
+
+	systable_endscan(tgscan);
+	heap_close(tgrel, AccessShareLock);
+
+	return tgoid;
+}
+
+/*
+ * Create an internal trigger that blocks INSERTs on a table.
+ *
+ * The blocking trigger is used to block accidental INSERTs on a hypertable's
+ * root table.
+ */
+static Oid
+add_insert_blocker_trigger(Oid relid)
+{
+	ObjectAddress objaddr;
+	char	   *relname = get_rel_name(relid);
+	Oid			schemaid = get_rel_namespace(relid);
+	char	   *schema = get_namespace_name(schemaid);
+	CreateTrigStmt stmt = {
+		.type = T_CreateTrigStmt,
+		.row = true,
+		.timing = TRIGGER_TYPE_BEFORE,
+		.trigname = INSERT_BLOCKER_NAME,	/* Note, internal triggers get the
+											 * OID appended to the name */
+		.relation = makeRangeVar(schema, relname, -1),
+		.funcname = list_make2(makeString(INTERNAL_SCHEMA_NAME), makeString(INSERT_BLOCKER_NAME)),
+		.args = NIL,
+		.events = TRIGGER_TYPE_INSERT,
+	};
+
+	objaddr.objectId = get_insert_blocker_trigger(relid);
+
+	/* Trigger already exists. Do nothing */
+	if (OidIsValid(objaddr.objectId))
+		return objaddr.objectId;
+
+	objaddr = CreateTrigger(&stmt, NULL, relid, InvalidOid, InvalidOid, InvalidOid, true);
+
+	if (!OidIsValid(objaddr.objectId))
+		elog(ERROR, "could not create insert blocker trigger");
+
+	return objaddr.objectId;
+}
+
+TS_FUNCTION_INFO_V1(hypertable_add_insert_blocker_trigger);
+
+/*
+ * This function is exposed to add the blocking trigger on legacy hypertables
+ * that don't have the trigger. We can't do it from SQL code, because internal
+ * triggers cannot be added from SQL.
+ *
+ * We also move any data in the root table to chunks, in case tuples have been
+ * accidentally inserted into the root table while the trigger didn't exist.
+ */
+Datum
+hypertable_add_insert_blocker_trigger(PG_FUNCTION_ARGS)
+{
+	Oid			relid = PG_GETARG_OID(0);
+	Cache	   *hcache = hypertable_cache_pin();
+
+	/*
+	Hypertable *ht;
+
+	  TODO: We'd like to be able migrate data here, but our metadata catalog
```

Comment:
Need to figure out how to handle data migration, if at all (see my other comment about this).